File: old_files/palindroms.py
MOD = int(1e9) + 7

def generate_combination(memo, strings, string_index, index, word, chosen_count):
    n = len(strings[string_index])

    if index >= n:
        return 0

    possible_count = generate_combination(
        memo,
        strings,
        string_index,
        index + 1,
        word ^ (1 << (ord(strings[string_index][index]) - ord('a'))),
        chosen_count + 1
    ) % MOD

    possible_count += generate_combination(
        memo,
        strings,
        string_index,
        index + 1,
        word,
        chosen_count
    ) % MOD

    if chosen_count > 0:
        possible_count += evaluate(
            memo,
            strings,
            string_index + 1,
            word
        ) % MOD

    return possible_count % MOD

def is_palindrome(word):
    bit_count = 0
    while word != 0:
        if (word & 1) == 1:
            bit_count += 1

        if bit_count > 1:
            return False

        word = word >> 1

    return True

def evaluate(memo, strings, string_index, word):
    n = len(strings)
    if string_index >= n:
        print(word)
        return 1 if is_palindrome(word) else 0

    if memo[string_index][word] != -1:
        return memo[string_index][word]

    memo[string_index][word] = generate_combination(
        memo,
        strings,
        string_index,
        0,
        word,
        0
    ) % MOD
    return memo[string_index][word]

def palindromic_strings(strings):
    n = len(strings)
    memo = [[-1 for _ in range(2 ** 26 + 1)] for _ in range(n)]
    return evaluate(memo, strings, 0, 0)

if __name__ == "__main__":
    q = int(input())
    for _ in range(q):
        n = int(input())
        A = [input() for _ in range(n)]
        result = palindromic_strings(A)
        print(result)

File: old_files/palindroms.cpp
#include <bits/stdc++.h>

using namespace std;

int evaluate(
    vector<vector<int>>& memo,
    vector<string>& strings,
    int stringIndex,
    int word
);

/*
 * Complete the 'palindromicStrings' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts STRING_ARRAY a as parameter.
 */
const int MOD = 1e9 + 7;


int generateCombination(
    vector<vector<int>>& memo,
    vector<string>& strings,
    int stringIndex,
    int index,
    int word,
    int chosenCount
) {
    int n = strings[stringIndex].size();

    if (index >= n) {
        return 0;
    }

    int possibleCount = generateCombination(
        memo,
        strings,
        stringIndex,
        index + 1,
        word ^ (1 << strings[stringIndex][index] - 'a'),
        chosenCount + 1
    );
    possibleCount %= MOD;

    possibleCount += generateCombination(
        memo,
        strings,
        stringIndex,
        index + 1,
        word,
        chosenCount
    );
    possibleCount %= MOD;

    if (chosenCount > 0) {
        possibleCount += evaluate(
            memo,
            strings,
            stringIndex + 1,
            word
        );
        possibleCount %= MOD;

    }

    return possibleCount % MOD;
}

bool isPalindrome(int word) {
    int bitCount = 0;
    while (word != 0) {
        if ((word & 1) == 1) {
            bitCount += 1;
        }

        if (bitCount > 1) {
            return false;
        }

        word = word >> 1;
    }

    return true;
}

int evaluate(
    vector<vector<int>>& memo,
    vector<string>& strings,
    int stringIndex,
    int word
) {
    int n = strings.size();
    if (stringIndex >= n) {
        if (isPalindrome(word)) {
            return 1;
        }

        return 0;
    }

    if (memo[stringIndex][word] != -1) {
        return memo[stringIndex][word];
    } 

    memo[stringIndex][word] = generateCombination(
        memo,
        strings,
        stringIndex,
        0,
        word,
        0
    );
    memo[stringIndex][word] %= MOD;
    return memo[stringIndex][word];
}

int palindromicStrings(vector<string> strings) {
    int n = strings.size();
    vector<vector<int>> memo(n, vector<int>(pow(2, 26) + 1, -1));
    printf("test\n");
    // return evaluate(memo, strings, 0, 0);
    return 1;
}


int main()
{
    int q, n, result;

    cin >> q;
    for (int i = 0; i < q; i++) {
        cin >> n;
        vector<string> A(n);
        for (int j = 0; j < n; j++) {
            cin >> A[j];
        }

        result = palindromicStrings(A);

        cout << result << endl;
    }

    return 0;
}
